Takes View Model- Common

KJoslyn opened the pull request:
Sister PR to jvm [#458](https://github.com/WycliffeAssociates/otter-jvm/pull/458)

KJoslyn committed:
Add slug to Book

KJoslyn committed:
Refactor Resource to have title and body Components

KJoslyn committed:
Add fields to Chunk and Workbook

KJoslyn committed:
Recordable

KJoslyn committed:
Rewrite RecordTake to support workbook pattern

KJoslyn committed:
Clean up Recordable

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -104,12 +113,27 @@ class WorkbookRepository(private val db: IDatabaseAccessors) : IWorkbookReposito
-        return Resource(
+        val bodyComponent = body?.let {
+            textItem(it)?.let { bodyTextItem ->
+                Resource.Component(
```
If the body text is null, should the body component be null or do we just return a null Resource?

aunger replied:
I don't know.

jsarabia replied:
why would you return a null resource? The resource could have a title?

KJoslyn replied:
I assume that if the resource has an associated body Content, but the body content is empty, there is something wrong with the resource. In this case, do we still return the resource with a null body? In general, I'm not sure what to do when things happen that we don't expect. We seem to have a precedent for returning null objects in that case.

jsarabia replied:
> In general, I'm not sure what to do when things happen that we don't expect. We seem to have a precedent for returning null objects in that case.

Well this is a problem we need to address; we need to be very deliberate about why we use nullable types, and where.

My understanding is that a Resource is a pairing of Title and Body. If this assumption is true AND body is optional, then it would follow that we need to return resources that have a null body, because in that case the null is not a "null" as in error but null as in "optional." My assumption is that this is true because some things have headers with no body (which is why we have some resource card things with only one status indicator).

If I'm incorrect in this assumption, which use case is this exactly? If "Resource" is not meant to have an optional body, then the body field should not be nullable, and we should be handling some kind of error case here, not propagating nulls through the app's operation.

KJoslyn replied:
You're correct, the body should be nullable since we may have headers with no body. I'm talking about the case where we are calling constructResource() to build a Resource, and the `body` parameter is not null BUT `textItem(body)` returns null. This means there is something wrong with the body Content object, i.e. either the `format` field of the Content object is incorrect or null, or the `text` field is null. In this case, we have found a faulty body Content, so I assume the entire Resource is corrupt. My question is then what do we do in this situation?

KJoslyn replied:
Update: when `textItem` gets called, the `content` parameter will be a `Content` with a `contentType` of `TEXT`, `TITLE`, or `BODY`. In all of these cases, `content.format` and `content.text` should be non-null. HOWEVER, up until this point, UsfmProjectReader had been creating `Content` objects with null `format` and `text` fields. I have since changed that.

HOWEVER (x2), ParseUsfm does not appear to be working correctly, so for now, there are verse `Content` objects with null text fields. I am leaving a TODO here to fix this function once ParseUsfm is fixed. Even so, I have made the return type of `textItem` non-nullable since we do not want null TextItems propagating through the app.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/Recordable.kt:
```diff
@@ -0,0 +1,54 @@
+        )
+    }
+
+    override fun equals(other: Any?): Boolean {
```
This is used so that we know whether the user selected a new Recordable, so we don't trigger onChange events by setting a Recordable to the same thing
that was selected before

jsarabia replied:
if you're overriding equals you probably need to override hashcode

KJoslyn replied:
Actually, overriding hashcode is a lot easier than I expected since I can just use `Objects.hash`

jsarabia replied:
This looks off to me. I've never written one myself but when I've seen articles and they all follow a similar looking pattern. I'd prefer to defer to the expertise in the Effective Java book which I believe I saw on Craig's desk. Could you take a look at it?

KJoslyn replied:
Objects.hash should work just fine. It uses the hashCode() implementation of each parameter that is passed to Objects.hash. For a little more info...
https://stackoverflow.com/questions/45832458/use-objects-hash-or-own-hashcode-implementation

KJoslyn replied:
I just realized that the only classes that implement Recordable are data classes, for which Kotlin already provides an implementation for `equals()` and `hashCode()`. 

While we can't really guarantee that data classes are the only classes that will ever implement Recordable, I personally no longer think that it's worth keeping custom equals() and hashcode() implementations in Recordable. If we do this, classes that implement Recordable still have to override equals() and hashcode() and call Recordable's default implementation of equals() or hashcode() in the function body. (And in that case, it's not a true equals() or hashcode() for the whole object since you're only using the properties that are part of the Recordable interface.)

If we want to guarantee that objects that implement Recordable are data classes, we would have to turn Recordable into a data class itself rather than an interface, and have Chunk and Resource.Component extend it, but this seems like overkill to me.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/Recordable.kt:
```diff
@@ -0,0 +1,54 @@
+            audio = chunk.audio,
+            start = chunk.start,
+            end = chunk.end
+        )
```
We may need to change this function or add another build function later for when a user is recording an entire chapter. I'm not sure how this will work.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/Recordable.kt:
```diff
@@ -0,0 +1,54 @@
+import org.wycliffeassociates.otter.common.data.workbook.*
+
+class Recordable private constructor(
+    val textItem: TextItem? = null,
```
This is nullable since Chunk's textItem is nullable (which in turn is because Content's text column is nullable.)
This feels bad, and it seems like we shouldn't even be able to create a Recordable if the text is null

aunger replied:
I think I agree.  Does it make sense to just make this textItem field non-nullable and have build() return a nullable?

KJoslyn replied:
Perhaps, but if we just return a null Recordable, we still wouldn't know what happened. Should we throw an exception, or perhaps use logging to log what happened? We don't seem to do much of either of these options in the current app.

jsarabia replied:
Let's talk about the purpose of this object; textItem may not make sense to be owned by it?

KJoslyn replied:
Recordable is meant to be a generalized object for something that can be recorded, be it scripture or resources. It should have the information needed to display the text to record, and to generate a wav file (filename and all) when the user records a take. That's my idea for it anyway

jsarabia replied:
I think it should probably be focused on just information just needed to create a recording, as you can see textItem already starts to introduce behavior that some recordables might have and others don't.

Recordable could be an interface maybe?

KJoslyn replied:
Hm, the concept of a Recordable that doesn't have a text item doesn't make sense to me- what are you recording then? I only originally made textItem nullable because it can propagate from Content's nullable `text` column (but I don't think we'd ever want to make a Recordable based on such a Content with a null text field.)

I'm not saying there definitely isn't a better way to do this, but if you take a look at how I'm building the Recordables in ResourcesViewModel and using them in TakesViewModel, I wonder if you would agree or disagree with the approach.

jsarabia replied:
> Hm, the concept of a Recordable that doesn't have a text item doesn't make sense to me- what are you recording then?

What is the purpose of the apps that we're building? Similarly:

> I don't think we'd ever want to make a Recordable based on such a Content with a null text field.

I do

KJoslyn replied:
Updating this thread after my conversation with Joe yesterday: 

Oral-to-oral translations are a case where the `text` field will be null. Drawing from the single responsibility principle, we are looking at creating an interface/class whose job is simply to name an audio file. The old `Recordable` object will either implement this interface or contain this "namer" object. Then we will handle the separate use cases of text-to-audio or audio-to-audio, perhaps by subclassing Recordable.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/RecordTake.kt:
```diff
@@ -29,142 +23,139 @@ class RecordTake(
+                ?.plus(1)
+                ?: 1
+        )
+    }
```
This is untested, but might be better:
```
    private fun getNewTakeNumber(recordable: Recordable): Single<Int> =
        recordable.audio.takes
            .map(Take::number)
            .reduce(0, Math::max)
            .map { it + 1 }
```

KJoslyn replied:
Not sure why, but this gets hung up during execution just like recordable.audio.takes.count() when I tried that

aunger replied:
Ohh!  It's because the relay never sends Complete.  There might be a better way of handling this, but this occurs to me: Replace `recordable.audio.takes` with `recordable.audio.takes.take(recordable.audio.takes.values.size.toLong())`

aunger replied:
That's not great either, though, is it? I wish there were a slicker way to get the buffered values from a relay, or at least its buffer size. What you had originally may be the best option.

KJoslyn replied:
Maybe this is better:
```
    private fun getNewTakeNumber(recordable: Recordable): Single<Int> =
        recordable.getAllTakesObservable()
            .map(Take::number)
            .reduce(0, Math::max)
            .map { it + 1 }
```

and in Recordable,

```
    fun getAllTakesObservable(): Observable<Take> = audio.takes.take(audio.takes.values.size.toLong())
```

`getNewTakeNumber()` probably shouldn't depend on the `takes` relay being specifically a ReplayRelay anyway

KJoslyn replied:
Or maybe we should really create an interface with a method to get all takes? Then either Recordable (or better yet, AssociatedAudio?) can implement this interface, and `getNewTakeNumber()` would take an object that implements the interface. @sarabiaj does this sound good?
```
class AssociatedAudio: I[SomeInterface] {
...
}

interface I[SomeInterface] {
    fun getAllTakesObservable()
    fun acceptTake()
}
```
Or perhaps we don't need this interface and we can just depend on AssociatedAudio- I might be taking the Open-Closed principle too far. I'm definitely still learning here

aunger replied:
I'd like us to add an extension method, `ReplayRelay.takeBuffered() = this.take(this.values.count)`.

I opened an [issue on RxRelay](https://github.com/JakeWharton/RxRelay/issues/52), and he can implement it in a more efficient way, but we should do it as an extension method in the meantime.

Then we can make the domain code do whatever.  Extension methods or interfaces as you're describing should be fine, but just making the original `RecordTake::getNewTakeNumber` use the new relay interface would feel pretty good to me.

What do you think?

KJoslyn replied:
Yeah I think that's a good idea!

aunger replied:
Well, I failed on my RxRelay change-request. 

We could just implement our own `ReplayRelay.takeBuffered()` extension method, even though it's a little inefficient.  I also think your `getAllTakesObservable` idea is workable. I also had some fun thinking about going further with an Observable that always has the next take number ready for you to use with `takes.scan(0) { acc, take -> Math.max(acc, take.number) }`, but that required miry time-based debouncing and/or caching, and it was hard not to make it break our lazy loading.

However, the more I think about your original implementation, the more I like it, and the more I regret raising a fuss. _Ctrl-Z?_ We really don't need to stay in the [reactive paradigm](http://introtorx.com/Content/v1.0.10621.0/10_LeavingTheMonad.html#WhyLeaveTheMonad) for this operation, so just calling getValues is fine.

If you do go back to the array, just be careful; your code `relay.getValues(arrayOfNulls<Take>(relay.values.size))` actually calls `getValues` (along with its array copy) twice.  Do `getValues(emptyArray())` instead.

KJoslyn replied:
Haha, ok let's _ctrl-z_ then! Thanks for looking into all of this anyway, I feel like the change-request had some potential and was a good idea.

Anyway, perhaps instead of `AssociatedAudio::getAllTakesObservable`, I will create an `AssociatedAudio::getAllTakes` method that simply returns and array of all the takes (and only calls `getValues` once, as you mentioned!)

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/Recordable.kt:
```diff
@@ -0,0 +1,54 @@
+                        && this.contentType == other.contentType
+            }
+        } ?: return false
+    }
```
I'm unsure but there seems to be unsafe casting here.  Maybe this:
```
    override fun equals(other: Any?): Boolean {
        return (other as? Recordable)?.let {
            this.textItem?.text == other.textItem?.text
                    && this.start == other.start
                    && this.end == other.end
                    && this.sort == other.sort
                    && this.contentType == other.contentType
        } ?: false
    }
```

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Workbook.kt:
```diff
@@ -2,5 +2,7 @@ package org.wycliffeassociates.otter.common.data.workbook
-    val target: Book
+    val target: Book,
+    val sourceLanguageSlug: String?,
+    val targetLanguageSlug: String?
```
Why are these nullable?

KJoslyn replied:
See line 49 of WorkbookRepository.kt. Collection's `resourceContainer` field is nullable, which is where we get the language slug. Should WorkbookRepository.get() just fail (throw an exception??) if we can't get the language slug due to a null `resourceContainer`?

jsarabia replied:
Null values are like checked exceptions in Java (and this is a good thing if handled right). If handled correctly at the first point where nulls could get introduced, then you don't have to worry about null values creeping into later parts of the app. If you do let them in (and kotlin makes it so you the programmer explicitly let null values enter your code, the language doesn't by default), at each point further down the chain you have to handle the null, and most of the time you're not going to know what to do because it's this "...how could something get this far along if it doesn't?" The null most of the time needs to be handled at the point it gets introduced, and from that point on, things shouldn't be nullable. There are some exceptions, where null values are used as a kind of "optional" type. But as a result, the "option" needs to be checked in all places it is considered.

Now; from a correctness standpoint, a translation project has a source (it has to, otherwise we have no idea any information like how many chapters, chunks, etc) and sources have to have resource contianers. We legitimately CANNOT move forward without a source. So a workbook can't be created without one. A workbook cannot be created without a target language either, that doesn't make sense.

So, if we have to have a source, and we have to have a target, those works both have to have languages.

Allowing them to be nulls is therefore by definition a bug (may not present itself, but why even allow for it?). 

What has a null resource container?

KJoslyn replied:
Very good points. 

I think no Collection should have a null RC. From inspecting the code again, we create Collections with a null resource container as an intermediate step during the import process, but immediately after they are mapped to CollectionEntities, we set the dublinCoreFk to what it should be before saving the entity to the database. (dublinCoreFk is non-nullable in the DB.) So, we really shouldn't be encountering a Collection with a null resource container that came from the database- if we encounter that, we should handle it immediately.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/RecordTake.kt:
```diff
@@ -29,142 +23,139 @@ class RecordTake(
-        } else {
-            "$verseFormat-$verseFormat".format(content.start, content.end)
-        }
+        val languageSlug = workbook.targetLanguageSlug ?: ""
```
change this if you make language non-null

KJoslyn committed:
PR comments

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Resource.kt:
```diff
@@ -1,9 +1,15 @@
+    val title: Component,
+    val body: Component?
+) {
+    class Component(
```
can this be a data class?

KJoslyn replied:
Sure!

KJoslyn committed:
PR Comment and first test

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/RecordTake.kt:
```diff
@@ -29,142 +23,139 @@ class RecordTake(
+                LaunchPlugin.Result.SUCCESS -> {
+                    if (take.file.length() == EMPTY_WAVE_FILE_SIZE) {
+                        take.file.delete()
+                        Single.just(Result.NO_AUDIO)
```
To simplify, you can remove these `Single.just` calls and use `map` instead of `flatMap`.

KJoslyn replied:
Great suggestion!

KJoslyn committed:
PR Comments

KJoslyn committed:
Test Cases for RecordTake

KJoslyn committed:
Updated TestWorkbookRepo

KJoslyn committed:
FileNamer and unit tests

KJoslyn committed:
WorkbookFileNamerBuilder + Recordable is now an interface

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Workbook.kt:
```diff
@@ -2,5 +2,7 @@ package org.wycliffeassociates.otter.common.data.workbook
-    val target: Book
+    val target: Book,
+    val sourceLanguageSlug: String,
+    val targetLanguageSlug: String
```
Should the language slugs be added to Book instead?

KJoslyn replied:
I believe I had asked @sarabiaj about this and we decided to keep it in Workbook (?)

jsarabia replied:
I have no opinion other than recognizing we need it

KJoslyn replied:
I like it better in Book, just made the change

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/WorkbookFileNamerBuilder.kt:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    private fun checkBuildLevel() {
+        if (buildLevel.ordinal != BuildLevel.values().map { it.ordinal }.max() ) {
```
Extract this as companion object value.  It doesn't need to be calculated more than once.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/WorkbookFileNamerBuilder.kt:
```diff
@@ -0,0 +1,97 @@
+
+    private var buildLevel: BuildLevel = BuildLevel.EMPTY
+
+    enum class BuildLevel {
```
private

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/WorkbookFileNamerBuilder.kt:
```diff
@@ -0,0 +1,97 @@
+                it.ordinal == buildLevel.ordinal - 1
+            }
+            fun min(first: BuildLevel, second: BuildLevel): BuildLevel {
+                return when (first.ordinal < second.ordinal) {
```
Most or all of the `.ordinal` uses in this file can be replaced by stuff from https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/index.html

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/WorkbookRepository.kt:
```diff
@@ -82,34 +91,45 @@ class WorkbookRepository(private val db: IDatabaseAccessors) : IWorkbookReposito
     }
 
     private fun constructResource(title: Content, body: Content?): Resource? {
         val titleTextItem = textItem(title)
```
Now that you've removed the null check here, it's cleaner to inline this instead of making a val.

KJoslyn committed:
UsfmProjectReader- put verse text and format into verse Content objects

However, parseUsfm appears to be broken, so the text will be incorrect for the verses

aunger commented on src/test/kotlin/org/wycliffeassociates/otter/common/data/workbook/TestAssociatedAudio.kt:
```diff
@@ -0,0 +1,64 @@
+                val audio = AssociatedAudio(createTakesRelay(it))
+                audio.getNewTakeNumber().blockingGet()
+            }
+    }
```
If I create a take, then delete it, what should the next take number _after that_ be?  Can you test this property?

KJoslyn replied:
The way this is implemented, deleted takes' take numbers are still treated as "used" take numbers. So, if we have 3 takes, and delete take 3, the next take we create would be take 4. In reality, maybe this should be the new take 3. But there are a few problems...
- First, we could be overwriting the old take 3 file if we use the same filename. We might not want that. (?) It depends on what the purpose of keeping around deleted takes is.
- Second, the filenames in the database (Take_entity) table would be the same for the old take 3 and the new take 3

Now, the take numbers that the _view_ displays to the user doesn't necessarily have to correspond to the take numbers in the filenames. But this can be confusing if they are ever aware of the filename of the take they are listening to. When they open, say, ocenaudio, to edit a take, you can see the take filename at the top.

Another more complicated case that I have been thinking about: If we have 5 takes and delete take 2, would we have to shift every take number down? This doesn't necessarily sound like a good idea to me since the user might have been listening to another take and associating it with a particular take number that could now be different. Eventually, we will have the waveform to display as well as the number, but I think it could still be confusing to change the take numbers.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/WorkbookFileNamerBuilder.kt:
```diff
@@ -0,0 +1,97 @@
+
+class WorkbookFileNamerBuilder: FileNamer.Builder() {
+
+    private var buildLevel: BuildLevel = BuildLevel.EMPTY
```
We talked about holding state in this object, and there may be a better way.

KJoslyn replied:
Yes, we are already essentially already holding state in the WorkbookViewModel, so we can just set everything all at once. We all agree on this now! :) I greatly simplified the FileNamer as a result.

KJoslyn committed:
Simplified WorkbookFileNameBuilder

KJoslyn committed:
Further simplified FileNamer

KJoslyn committed:
Minor changes

KJoslyn committed:
Remove equals() and hashCode() from Recordable

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/WorkbookFileNamerBuilder.kt:
```diff
@@ -0,0 +1,26 @@
+    ) = FileNamer(
+        bookSlug = workbook.target.slug,
+        languageSlug = workbook.targetLanguageSlug,
+        chapterCount = workbook.target.chapters.count().subscribeOn(Schedulers.io()).blockingGet(),
```
not sure what I think of moving to an io thread and doing a blocking get especially if you're not wrapping the return value here in an observable and doing this async

KJoslyn replied:
Yeah I think it's fine on the UI thread actually- it's a small operation that shouldn't cause any issues

KJoslyn committed:
PR Comments

KJoslyn committed:
Merge branch 'dev' into kj-takes-tabpane-newtake

KJoslyn committed:
Fix broken test case

Commit:
Merge branch 'dev' into kj-takes-tabpane-newtake
